hudi-spark-datasource/hudi-spark3/src/test/scala/org/apache/hudi/spark3/internal/TestHoodieDataSourceInternalBatchWrite.java

```diff
@@ -0,0 +1,264 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hudi.spark3.internal;
+
+import org.apache.hudi.common.testutils.HoodieTestDataGenerator;
+import org.apache.hudi.config.HoodieWriteConfig;
+import org.apache.hudi.internal.HoodieDataSourceInternalWriterTestBase;
+import org.apache.hudi.table.HoodieSparkTable;
+import org.apache.hudi.table.HoodieTable;
+import org.apache.hudi.testutils.HoodieClientTestUtils;
+
+import org.apache.spark.sql.Dataset;
+import org.apache.spark.sql.Row;
+import org.apache.spark.sql.catalyst.InternalRow;
+import org.apache.spark.sql.connector.write.DataWriter;
+import org.junit.jupiter.api.Test;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+import static org.apache.hudi.testutils.SparkDatasetTestUtils.ENCODER;
+import static org.apache.hudi.testutils.SparkDatasetTestUtils.STRUCT_TYPE;
+import static org.apache.hudi.testutils.SparkDatasetTestUtils.getConfigBuilder;
+import static org.apache.hudi.testutils.SparkDatasetTestUtils.getRandomRows;
+import static org.apache.hudi.testutils.SparkDatasetTestUtils.toInternalRows;
+
+/**
+ * Unit tests {@link HoodieDataSourceInternalBatchWrite}.
+ */
+public class TestHoodieDataSourceInternalBatchWrite extends
+    HoodieDataSourceInternalWriterTestBase {
+
+  @Test
+  public void testDataSourceWriter() throws Exception {
+    // init config and table
+    HoodieWriteConfig cfg = getConfigBuilder(basePath).build();
+    HoodieTable table = HoodieSparkTable.create(cfg, context, metaClient);
+    String instantTime = "001";
+    // init writer
+    HoodieDataSourceInternalBatchWrite dataSourceInternalBatchWrite =
+        new HoodieDataSourceInternalBatchWrite(instantTime, cfg, STRUCT_TYPE, sqlContext.sparkSession(), hadoopConf);
+    DataWriter<InternalRow> writer = dataSourceInternalBatchWrite.createBatchWriterFactory(null).createWriter(0, RANDOM.nextLong());
+
+    String[] partitionPaths = HoodieTestDataGenerator.DEFAULT_PARTITION_PATHS;
```

Comment:
again I see some opportunity for code reuse here. Do you think we can avoid duplicating the test code across two test classes. Except for few lines, rest could be made common.

Reply:
> Few more comments. thanks for code reuse. now most of our core logic is in one place for both datasources.
> btw, can you add tests to HoodieSparkSqlWriterSuite as well.
> You can check out previous tests [here](https://github.com/apache/hudi/pull/1834).

Thanks for the quick review. 
I am not very sure about what kind of test do you want to add here.
The test ```test("test bulk insert dataset with datasource impl")``` in ```HoodieSparkSqlWriterSuite``` can also be used for testing this pr. You can do this through ```mvn test xxx -Pspark3```.

Reply:
Yea I can try to reduce the duplicated code between ```TestHoodieBulkInsertDataInternalWriter``` and ```TestHoodieDataSourceInternalBatchWrite```.

The code of ```org.apache.hudi.internal.TestHoodieBulkInsertDataInternalWriter``` is almost the same as  ```org.apache.hudi.spark3.internal.TestHoodieBulkInsertDataInternalWriter```. However it's kind of tricky to avoid this. Because one ```HoodieBulkInsertDataInternalWriter``` implements  ```org.apache.spark.sql.connector.write.DataWriter``` which is only compatible with Spark 3 but another ```HoodieBulkInsertDataInternalWriter``` implements ```org.apache.spark.sql.sources.v2.writer.DataWriter``` which is only compatible with Spark 2.

Reflection is one way to solve this problem.

Reply:
if we can solve it using Reflection, it might be okay for a test.